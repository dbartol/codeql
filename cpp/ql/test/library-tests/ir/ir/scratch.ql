import cpp

string getQName(Function f) {
  result = f.getQualifiedName() or
  not exists(f.getQualifiedName()) and result = "??"
}

from Function func
select func, getQName(func), func.getDeclaringType().getUnspecifiedType()