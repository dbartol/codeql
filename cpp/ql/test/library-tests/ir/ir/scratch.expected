| bad_asts.cpp:5:10:5:10 | operator= | Bad::S::operator= | bad_asts.cpp:5:10:5:10 | S |
| bad_asts.cpp:5:10:5:10 | operator= | Bad::S::operator= | bad_asts.cpp:5:10:5:10 | S |
| bad_asts.cpp:9:9:9:22 | MemberFunction | Bad::S::MemberFunction | bad_asts.cpp:5:10:5:10 | S |
| bad_asts.cpp:19:10:19:10 | Point | Bad::Point::Point | bad_asts.cpp:19:10:19:14 | Point |
| bad_asts.cpp:19:10:19:10 | Point | Bad::Point::Point | bad_asts.cpp:19:10:19:14 | Point |
| bad_asts.cpp:19:10:19:10 | operator= | Bad::Point::operator= | bad_asts.cpp:19:10:19:14 | Point |
| bad_asts.cpp:19:10:19:10 | operator= | Bad::Point::operator= | bad_asts.cpp:19:10:19:14 | Point |
| bad_asts.cpp:22:5:22:9 | Point | Bad::Point::Point | bad_asts.cpp:19:10:19:14 | Point |
| file://:0:0:0:0 | operator= | __va_list_tag::operator= | file://:0:0:0:0 | __va_list_tag |
| file://:0:0:0:0 | operator= | __va_list_tag::operator= | file://:0:0:0:0 | __va_list_tag |
| ir.cpp:412:8:412:8 | operator= | Point::operator= | ir.cpp:412:8:412:12 | Point |
| ir.cpp:412:8:412:8 | operator= | Point::operator= | ir.cpp:412:8:412:12 | Point |
| ir.cpp:417:8:417:8 | operator= | Rect::operator= | ir.cpp:417:8:417:11 | Rect |
| ir.cpp:417:8:417:8 | operator= | Rect::operator= | ir.cpp:417:8:417:11 | Rect |
| ir.cpp:525:7:525:7 | operator= | U::operator= | ir.cpp:525:7:525:7 | U |
| ir.cpp:525:7:525:7 | operator= | U::operator= | ir.cpp:525:7:525:7 | U |
| ir.cpp:599:5:599:10 | String | String::String | ir.cpp:597:8:597:13 | String |
| ir.cpp:600:5:600:10 | String | String::String | ir.cpp:597:8:597:13 | String |
| ir.cpp:601:5:601:10 | String | String::String | ir.cpp:597:8:597:13 | String |
| ir.cpp:602:5:602:11 | ~String | String::~String | ir.cpp:597:8:597:13 | String |
| ir.cpp:604:13:604:21 | operator= | String::operator= | ir.cpp:597:8:597:13 | String |
| ir.cpp:605:13:605:21 | operator= | String::operator= | ir.cpp:597:8:597:13 | String |
| ir.cpp:607:17:607:21 | c_str | String::c_str | ir.cpp:597:8:597:13 | String |
| ir.cpp:628:7:628:7 | C | C::C | ir.cpp:628:7:628:7 | C |
| ir.cpp:628:7:628:7 | C | C::C | ir.cpp:628:7:628:7 | C |
| ir.cpp:628:7:628:7 | operator= | C::operator= | ir.cpp:628:7:628:7 | C |
| ir.cpp:628:7:628:7 | operator= | C::operator= | ir.cpp:628:7:628:7 | C |
| ir.cpp:628:7:628:7 | ~C | C::~C | ir.cpp:628:7:628:7 | C |
| ir.cpp:630:16:630:35 | StaticMemberFunction | C::StaticMemberFunction | ir.cpp:628:7:628:7 | C |
| ir.cpp:634:9:634:30 | InstanceMemberFunction | C::InstanceMemberFunction | ir.cpp:628:7:628:7 | C |
| ir.cpp:638:17:638:37 | VirtualMemberFunction | C::VirtualMemberFunction | ir.cpp:628:7:628:7 | C |
| ir.cpp:642:10:642:20 | FieldAccess | C::FieldAccess | ir.cpp:628:7:628:7 | C |
| ir.cpp:652:10:652:20 | MethodCalls | C::MethodCalls | ir.cpp:628:7:628:7 | C |
| ir.cpp:658:5:658:5 | C | C::C | ir.cpp:628:7:628:7 | C |
| ir.cpp:713:8:713:8 | operator= | Outer<long>::operator= | ir.cpp:713:8:713:12 | Outer<long> |
| ir.cpp:713:8:713:8 | operator= | Outer<long>::operator= | ir.cpp:713:8:713:12 | Outer<long> |
| ir.cpp:715:12:715:12 | Func | Outer<long>::Func | ir.cpp:713:8:713:12 | Outer<long> |
| ir.cpp:715:12:715:15 | Func | Outer<T>::Func | ir.cpp:713:8:713:12 | Outer<T> |
| ir.cpp:715:12:715:15 | Func | Outer<long>::Func | ir.cpp:713:8:713:12 | Outer<long> |
| ir.cpp:745:8:745:8 | Base | Base::Base | ir.cpp:745:8:745:11 | Base |
| ir.cpp:745:8:745:8 | operator= | Base::operator= | ir.cpp:745:8:745:11 | Base |
| ir.cpp:748:3:748:6 | Base | Base::Base | ir.cpp:745:8:745:11 | Base |
| ir.cpp:750:3:750:7 | ~Base | Base::~Base | ir.cpp:745:8:745:11 | Base |
| ir.cpp:754:8:754:8 | Middle | Middle::Middle | ir.cpp:754:8:754:13 | Middle |
| ir.cpp:754:8:754:8 | operator= | Middle::operator= | ir.cpp:754:8:754:13 | Middle |
| ir.cpp:757:3:757:8 | Middle | Middle::Middle | ir.cpp:754:8:754:13 | Middle |
| ir.cpp:759:3:759:9 | ~Middle | Middle::~Middle | ir.cpp:754:8:754:13 | Middle |
| ir.cpp:763:8:763:8 | Derived | Derived::Derived | ir.cpp:763:8:763:14 | Derived |
| ir.cpp:763:8:763:8 | operator= | Derived::operator= | ir.cpp:763:8:763:14 | Derived |
| ir.cpp:766:3:766:9 | Derived | Derived::Derived | ir.cpp:763:8:763:14 | Derived |
| ir.cpp:768:3:768:10 | ~Derived | Derived::~Derived | ir.cpp:763:8:763:14 | Derived |
| ir.cpp:772:8:772:8 | MiddleVB1 | MiddleVB1::MiddleVB1 | ir.cpp:772:8:772:16 | MiddleVB1 |
| ir.cpp:772:8:772:8 | operator= | MiddleVB1::operator= | ir.cpp:772:8:772:16 | MiddleVB1 |
| ir.cpp:775:3:775:11 | MiddleVB1 | MiddleVB1::MiddleVB1 | ir.cpp:772:8:772:16 | MiddleVB1 |
| ir.cpp:777:3:777:12 | ~MiddleVB1 | MiddleVB1::~MiddleVB1 | ir.cpp:772:8:772:16 | MiddleVB1 |
| ir.cpp:781:8:781:8 | MiddleVB2 | MiddleVB2::MiddleVB2 | ir.cpp:781:8:781:16 | MiddleVB2 |
| ir.cpp:781:8:781:8 | operator= | MiddleVB2::operator= | ir.cpp:781:8:781:16 | MiddleVB2 |
| ir.cpp:784:3:784:11 | MiddleVB2 | MiddleVB2::MiddleVB2 | ir.cpp:781:8:781:16 | MiddleVB2 |
| ir.cpp:786:3:786:12 | ~MiddleVB2 | MiddleVB2::~MiddleVB2 | ir.cpp:781:8:781:16 | MiddleVB2 |
| ir.cpp:790:8:790:8 | DerivedVB | DerivedVB::DerivedVB | ir.cpp:790:8:790:16 | DerivedVB |
| ir.cpp:790:8:790:8 | operator= | DerivedVB::operator= | ir.cpp:790:8:790:16 | DerivedVB |
| ir.cpp:793:3:793:11 | DerivedVB | DerivedVB::DerivedVB | ir.cpp:790:8:790:16 | DerivedVB |
| ir.cpp:795:3:795:12 | ~DerivedVB | DerivedVB::~DerivedVB | ir.cpp:790:8:790:16 | DerivedVB |
| ir.cpp:842:8:842:8 | PolymorphicBase | PolymorphicBase::PolymorphicBase | ir.cpp:842:8:842:22 | PolymorphicBase |
| ir.cpp:842:8:842:8 | PolymorphicBase | PolymorphicBase::PolymorphicBase | ir.cpp:842:8:842:22 | PolymorphicBase |
| ir.cpp:842:8:842:8 | operator= | PolymorphicBase::operator= | ir.cpp:842:8:842:22 | PolymorphicBase |
| ir.cpp:843:11:843:26 | ~PolymorphicBase | PolymorphicBase::~PolymorphicBase | ir.cpp:842:8:842:22 | PolymorphicBase |
| ir.cpp:846:8:846:8 | PolymorphicDerived | PolymorphicDerived::PolymorphicDerived | ir.cpp:846:8:846:25 | PolymorphicDerived |
| ir.cpp:846:8:846:8 | PolymorphicDerived | PolymorphicDerived::PolymorphicDerived | ir.cpp:846:8:846:25 | PolymorphicDerived |
| ir.cpp:846:8:846:8 | PolymorphicDerived | PolymorphicDerived::PolymorphicDerived | ir.cpp:846:8:846:25 | PolymorphicDerived |
| ir.cpp:846:8:846:8 | operator= | PolymorphicDerived::operator= | ir.cpp:846:8:846:25 | PolymorphicDerived |
| ir.cpp:846:8:846:8 | operator= | PolymorphicDerived::operator= | ir.cpp:846:8:846:25 | PolymorphicDerived |
| ir.cpp:846:8:846:8 | ~PolymorphicDerived | PolymorphicDerived::~PolymorphicDerived | ir.cpp:846:8:846:25 | PolymorphicDerived |
| ir.cpp:867:1:867:14 | String | String::String | ir.cpp:597:8:597:13 | String |
| ir.cpp:924:8:924:8 | operator= | SizedDealloc::operator= | ir.cpp:924:8:924:19 | SizedDealloc |
| ir.cpp:924:8:924:8 | operator= | SizedDealloc::operator= | ir.cpp:924:8:924:19 | SizedDealloc |
| ir.cpp:926:9:926:20 | operator new | SizedDealloc::operator new | ir.cpp:924:8:924:19 | SizedDealloc |
| ir.cpp:927:9:927:22 | operator new[] | SizedDealloc::operator new[] | ir.cpp:924:8:924:19 | SizedDealloc |
| ir.cpp:928:8:928:22 | operator delete | SizedDealloc::operator delete | ir.cpp:924:8:924:19 | SizedDealloc |
| ir.cpp:929:8:929:24 | operator delete[] | SizedDealloc::operator delete[] | ir.cpp:924:8:924:19 | SizedDealloc |
| ir.cpp:932:21:932:21 | operator= | Overaligned::operator= | ir.cpp:932:21:932:31 | Overaligned |
| ir.cpp:932:21:932:21 | operator= | Overaligned::operator= | ir.cpp:932:21:932:31 | Overaligned |
| ir.cpp:936:8:936:8 | DefaultCtorWithDefaultParam | DefaultCtorWithDefaultParam::DefaultCtorWithDefaultParam | ir.cpp:936:8:936:34 | DefaultCtorWithDefaultParam |
| ir.cpp:936:8:936:8 | DefaultCtorWithDefaultParam | DefaultCtorWithDefaultParam::DefaultCtorWithDefaultParam | ir.cpp:936:8:936:34 | DefaultCtorWithDefaultParam |
| ir.cpp:936:8:936:8 | operator= | DefaultCtorWithDefaultParam::operator= | ir.cpp:936:8:936:34 | DefaultCtorWithDefaultParam |
| ir.cpp:936:8:936:8 | operator= | DefaultCtorWithDefaultParam::operator= | ir.cpp:936:8:936:34 | DefaultCtorWithDefaultParam |
| ir.cpp:937:3:937:29 | DefaultCtorWithDefaultParam | DefaultCtorWithDefaultParam::DefaultCtorWithDefaultParam | ir.cpp:936:8:936:34 | DefaultCtorWithDefaultParam |
| ir.cpp:1048:12:1048:12 | (constructor) | ?? | file://:0:0:0:0 | decltype([...](...){...}) |
| ir.cpp:1048:12:1048:12 | (constructor) | ?? | file://:0:0:0:0 | decltype([...](...){...}) |
| ir.cpp:1048:12:1048:12 | (constructor) | ?? | file://:0:0:0:0 | decltype([...](...){...}) |
| ir.cpp:1048:12:1048:12 | operator= | ?? | file://:0:0:0:0 | decltype([...](...){...}) |
| ir.cpp:1048:15:1048:15 | operator() | ?? | file://:0:0:0:0 | decltype([...](...){...}) |
